packages/meditrak-server/src/routes/importSurveyResponses/assertCanImportSurveyResponses.js

```diff
@@ -57,3 +57,29 @@ export const assertCanImportSurveyResponses = async (
 
   return true;
 };
+
+export const assertSurveyResponseBatchPermissions = async (
+  accessPolicy,
+  models,
+  surveyResponses,
+) => {
+  // Assumes the data has already been validated
+  const entitiesGroupedBySurveyName = {};
+
+  for (const response of surveyResponses) {
+    // There are three valid ways to refer to the entity in a batch change
+    const entityCode =
+      response.entity_code ||
+      (response.entity_id
+        ? (await models.entity.findById(response.entity_id)).code
+        : (await models.facility.findById(response.clinic_id)).code);
+    const surveyName = (await models.survey.findById(response.survey_id)).name;
```

Comment:
A consideration here is that we will always perform 1 or 2 db queries per survey response

1. Potentially one for the entity (if no `entity_code` is provided)
2. Definitely one for the survey name

But it may be that eg 20 survey responses are submitted for a single survey and entity, which would make most of those queries redundant.

Overcoming this would require a more elaborate algorithm where we find all distinct entity/clinic/survey ids in the responses first, fetch the records next (3 queries max, for the whole batch) and then transform them in appropriate dictionaries to be used to eventually create `entitiesBySurveyName`.

I understand that this makes code more complicated and they say that premature optimisation is the root of all evil, so let's leave this as is. You may still want to attempt an compromise by pre-fetching distinct surveys, since this wouldn't be too complicated. `reduceToDictionary` is a useful util to then map survey ids to names.

Reply:
Fair enough, I've added pre-fetching of unique surveys for now